### PR TITLE
Target main and master in the ruleset, and warn if a repo has one

### PR DESCRIPTION
**Pull request by mikelward**
## The backdoor

A ruleset targeting only `~DEFAULT_BRANCH` (or only `main`) leaves a branch
literally named `master` unprotected: someone could push or merge straight
into it with none of the required checks applied. This just bit
mikelward/web, whose default branch was renamed `master` → `main` while
workflows kept pushing to the dead name.

## What `repo-rules` now does

A newly **created** ruleset's `conditions.ref_name.include` now names three
refs instead of one: `~DEFAULT_BRANCH`, `refs/heads/main`, and
`refs/heads/master`. `compute_scope`'s create-path fallback is kept in sync
so the merge-method conflict scan checks the real scope being written, not
the old single-ref one — a test confirms a squash-only ruleset scoped only
to `refs/heads/master` is now caught as a conflict, which it wasn't before.

An **update** to an existing ruleset still leaves its targeting alone, same
as every other field this script doesn't manage on that path (bypass
actors, enforcement, etc.) — see the updated header comment. That's what
`repo-rules-audit` is for.

**Also fixes a rename TOCTOU on the update path**, found during the Python
port's review (mikelward/repo#6, commit 5fe3c90, which fixed the same bug
there — no Codex thread on this repo, since Codex didn't flag it here).
`check_ruleset_ownership`'s re-check right before the write re-inspects id
`$EXISTING`'s enforcement and rule types, which catches someone editing
what that id *is* while the confirmation prompt was open, but not someone
*renaming* `'$RULESET'` away from it (or renaming a different ruleset onto
it) in that same window — the PUT addresses the ruleset by id, never by
name, so it would silently write to whatever ruleset now holds that id
while claiming to have updated the one named `'$RULESET'`. Now re-resolves
the name immediately before the write and refuses on any mismatch, fail
closed, mirroring the create path's own duplicate re-check just below it.

## What `repo-rules-audit` now checks

Two new, independent checks:

1. **Targeting completeness** (whenever the audited branch actually *is*
   the repository's real default — determined directly, not by whether
   `--branch` was given, so an explicit `--branch main` naming the real
   default still gets this check): for every ruleset that plainly covers
   it, does its `conditions.ref_name.include` contain all three refs — and
   does `conditions.ref_name.exclude` NOT literally carve one of them back
   out? An include list alone isn't the whole story: a ruleset that
   includes all three but excludes `refs/heads/master` leaves master exactly
   as unprotected as if it had never been listed. Missing/excluded refs are
   reported per ruleset. Same literal-match discipline as the existing
   bypass-actor scan — `~ALL` counts as covering all three (unless excluded).
   Literal coverage is checked *first*: a ruleset that already spells out
   all three refs literally is reported complete regardless of what other
   glob also sits in `include` (release-branch patterns and the like) —
   the glob only matters as a fallback for a ref that's genuinely not
   covered by any literal entry, and in that case the ruleset is reported
   unevaluated (a glob in `exclude` can never actually reach this check —
   the branch-coverage verdict computed earlier already routes it to
   unevaluated first). Either an unevaluated ruleset within `covers)`
   (a required ref genuinely behind a glob) *or* the branch-coverage
   verdict itself landing on `unevaluated` (an `include` that's only a
   glob, or a glob `exclude` over an otherwise-complete `include`)
   suppresses the `[ok] every ruleset ... targets ...` summary line —
   printing it would be a false all-clear over an audit that was never
   actually completed for that ruleset, whichever of the two paths it
   came from.
2. **A leftover `master` branch** (`GET /repos/{owner}/{repo}/branches/master`,
   `HTTP 404` in the ordinary case): independent of any ruleset, and
   regardless of which branch is being audited — deleting the branch
   removes the backdoor outright, rather than merely fencing it in.

## Tests

`repo-rules_test`: three new tests — the three refs present on create, a
squash-only ruleset scoped only to `refs/heads/master` now caught by the
conflict scan, and the rename-TOCTOU fix (both directions: a mismatched
second lookup is refused, an unchanged one still writes normally).

`repo-rules-audit_test`: new tests covering both directions of the
targeting check (clean with all three refs present; a gap naming the
missing ones), the `~ALL`/glob edge cases (including all-three-literal-
plus-an-unrelated-glob staying `ok`, and both the inner and outer
unevaluated paths correctly suppressing that same `ok` line), both
directions of the exclude fix (excluded ref still reported missing,
including via `~ALL`; an unrelated exclude entry stays clean), the
explicit-`--branch`-naming-the-real-default case, and both directions of
the master-branch check. Two pre-existing fixtures that used a single-ref
`["~DEFAULT_BRANCH"]` scope (testing bypass-actor behavior, unrelated to
targeting) were widened to the full three-ref set so they stay clean cases
for what they actually test.

`make test`: full suite green. `shellcheck -x` clean on all four changed
files.

## Review

Three rounds of Codex review, five findings total, all fixed:
excluded-ref false-clear, `ASKED_DEFAULT`-vs-real-default, literal-coverage
ordering ahead of a glob fallback, a contradictory `[ok]` over an
unevaluated ruleset (inner case), and the same contradiction via a second,
separate outer-verdict-unevaluated path.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01AoHridGxxapL68Y8fLwyRx

### CODE REVIEW

File: repo-rules-audit [hunk 1]
```diff
@@ -343,7 +358,49 @@ if gh api --paginate "repos/$REPO/rulesets?includes_parents=true" \
...
+                # GitHub's ref matching to find out (same discipline as the
+                # branch-coverage check above) -- so a literal-plus-glob
+                # include list is reported unevaluated rather than guessed
+                # at either way, alongside the other unevaluated rulesets.
+                if test "$ASKED_DEFAULT" -eq 1; then
+                    result=$(jq -r '
+                        def is_literal: test("[*?\\[]") | not;
+                        (.conditions.ref_name.include // []) as $inc
+                        | if ($inc | index("~ALL")) then "complete|"
+                          elif ($inc | any(is_literal | not)) then "unevaluated|"
```
[the remark this answers is not in the document]
Comment (mikelward) [reply repo-rules-audit]: Confirmed real — the targeting check only ever inspected `conditions.ref_name.include`. Fixed in b7124ae: it now also checks `conditions.ref_name.exclude` for each of the three wanted refs (same treatment whether reached via a plain literal include or via `~ALL`), and reports the excluded ref as missing rather than complete. A glob in `exclude` doesn't need separate handling here — the branch-coverage verdict computed earlier in the loop already routes any ruleset with a glob in `exclude` to `unevaluated` before this check ever runs, so `$exc` arriving here is guaranteed literal-only (documented in a comment). Added regression tests both directions: an include-all-three-but-exclude-master ruleset (plain and via `~ALL`) is a gap; an unrelated literal exclude entry stays clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: repo-rules-audit [hunk 2]
```diff
@@ -343,7 +358,49 @@ if gh api --paginate "repos/$REPO/rulesets?includes_parents=true" \
...
+                # ruleset scoped to some other branch (release, say) has any
+                # reason to carry. `~ALL` covers every one of the three
+                # unconditionally, by definition of the token -- not a gap.
+                # Any OTHER glob entry is different: it might cover main or
+                # master, or might not, and this script does not reimplement
+                # GitHub's ref matching to find out (same discipline as the
+                # branch-coverage check above) -- so a literal-plus-glob
+                # include list is reported unevaluated rather than guessed
+                # at either way, alongside the other unevaluated rulesets.
+                if test "$ASKED_DEFAULT" -eq 1; then
```
[the remark this answers is not in the document]
Comment (mikelward) [reply repo-rules-audit]: Confirmed real — `ASKED_DEFAULT` only means "no `--branch` was given," which isn't the same question as "is the audited branch the repository's real default." Fixed in b7124ae: both gates (the per-ruleset check and the final ok/gap report) now use `default_branch_matches` instead — the same function already used elsewhere in this script to resolve `~DEFAULT_BRANCH` in ruleset conditions, so it answers the right question regardless of how the branch was named, with no new API call on the already-common `ASKED_DEFAULT=1` path (it short-circuits there). Added a regression test for `--branch main` explicitly naming the real default.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: repo-rules-audit [hunk 3]
```diff
@@ -343,7 +358,70 @@ if gh api --paginate "repos/$REPO/rulesets?includes_parents=true" \
...
+                        # `$inc | index(.)` -- piping into $inc first rebinds
+                        # `.` to $inc itself before index() ever sees the
+                        # wanted ref, so every entry came back "found"
+                        # regardless of $inc.
+                        | if ($inc | index("~ALL")) then
+                              (["~DEFAULT_BRANCH", "refs/heads/main", "refs/heads/master"]
+                               | map(select(. as $want | ($exc | index($want)) != null))) as $miss
+                              | if ($miss | length) == 0 then "complete|"
+                                else "missing|" + ($miss | join(", ")) end
+                          elif ($inc | any(is_literal | not)) then "unevaluated|"
```
[the remark this answers is not in the document]
Comment (mikelward) [reply repo-rules-audit]: Confirmed real. Fixed in 1d24c4f: the jq now checks whether the three wanted refs are literally covered *first*, and only falls back to the glob check when a required ref genuinely isn't — so a ruleset that already spells out all three refs literally is reported complete regardless of what other patterns also sit in `include`. Added a regression test: all three literals plus an unrelated `refs/heads/release-*` glob is `ok`, not `[CHECK]`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: repo-rules-audit [hunk 4]
```diff
@@ -361,9 +439,32 @@ if gh api --paginate "repos/$REPO/rulesets?includes_parents=true" \
         echo "  [CHECK] rulesets with a pattern this script did not evaluate -- check by hand:"
         cat "$WORK/unevaluated"
     fi
+    if default_branch_matches; then
+        if test -s "$WORK/undertargeted"; then
+            gap "a ruleset covering $BRANCH does not target all of the default branch, refs/heads/main and refs/heads/master:"
+            cat "$WORK/undertargeted"
+        else
+            ok "every ruleset covering $BRANCH targets the default branch, refs/heads/main and refs/heads/master"
```
[the remark this answers is not in the document]
Comment (mikelward) [reply repo-rules-audit:484]: Confirmed real. Fixed in 1d24c4f: unevaluated-for-targeting rulesets are now tracked in their own file (`$WORK/targeting_unevaluated`), separately from `undertargeted`, and the final report has a third branch — neither `[ok]` nor `[GAP]`, just silent — while any remain, since the `[CHECK]` section already named them for manual review. Extended the existing mixed-pattern test (literal default-branch match + unrelated glob, main/master genuinely uncovered) to assert the `[ok]` summary line is absent.

As you note, fixing finding 1 first shrinks how often this state is even reached (a ruleset with all three refs literal-covered no longer goes through the unevaluated path at all), but a case with a genuinely-uncovered ref behind a glob still needs its own guard, which this adds.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: repo-rules-audit [hunk 5]
```diff
@@ -361,9 +455,39 @@ if gh api --paginate "repos/$REPO/rulesets?includes_parents=true" \
         echo "  [CHECK] rulesets with a pattern this script did not evaluate -- check by hand:"
         cat "$WORK/unevaluated"
     fi
+    if default_branch_matches; then
+        if test -s "$WORK/undertargeted"; then
+            gap "a ruleset covering $BRANCH does not target all of the default branch, refs/heads/main and refs/heads/master:"
+            cat "$WORK/undertargeted"
+        elif test -s "$WORK/targeting_unevaluated"; then
```
[the remark this answers is not in the document]
Comment (mikelward) [reply repo-rules-audit:476]: Confirmed real — a separate path from the two `$WORK/targeting_unevaluated` cases the last round added. Fixed in 90e14b9: the outer `unevaluated)` case (branch-coverage itself undetermined — a glob-only `include`, or an otherwise-complete `include` paired with a glob `exclude`) now also appends to `$WORK/targeting_unevaluated`, alongside its existing append to the shared `$WORK/unevaluated` bucket. Same gate at print time picks it up, so no change needed there. Added a regression test distinct from the existing mixed-pattern one: `include: ["refs/heads/release-*"]` with no bypass actors, asserting the `[CHECK]` line appears and the `[ok]` targeting summary does not.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: repo-rules-audit [hunk 6]
```diff
@@ -343,11 +359,103 @@ if gh api --paginate "repos/$REPO/rulesets?includes_parents=true" \
...
+                # GitHub's ref matching to find out (same discipline as the
+                # branch-coverage check above) -- so THAT case, and only
+                # that case, is reported unevaluated rather than guessed at
+                # either way, alongside the other unevaluated rulesets.
+                if default_branch_matches; then
+                    result=$(jq -r '
+                        def is_literal: test("[*?\\[]") | not;
+                        (.conditions.ref_name.include // []) as $inc
+                        | (.conditions.ref_name.exclude // []) as $exc
+                        | ["~DEFAULT_BRANCH", "refs/heads/main", "refs/heads/master"] as $want
```
[the remark this answers is not in the document]
Comment (mikelward) [reply repo-rules-audit:402]: Declining — this is a false positive against the PR's own stated design intent, not a bug.

`repo-rules`'s header comment (lines 29–39) is explicit about why it composes three refs instead of the current default's literal name: `~DEFAULT_BRANCH` "together with the literal `refs/heads/main` and `refs/heads/master`, **not the default branch alone**." The whole reason for the symbolic token is that GitHub re-resolves it automatically on every future rename — a ruleset carrying it keeps tracking the default branch with zero action required. The literal `main`/`master` entries are an *additive* safety net for those two specific legacy names, not a substitute for that tracking mechanism. `repo-rules-audit`'s own header (lines 44–53) frames the gap this check exists to catch as exactly "a ruleset created before repo-rules widened its targeting only ever covers `~DEFAULT_BRANCH`" — i.e. it's verifying the ruleset carries repo-rules' actual three-part convention, not just that it happens to cover today's known branch names.

If the audit accepted a literal ref matching the *current* default as equivalent to `~DEFAULT_BRANCH`, a ruleset scoped to literal `main`+`master` only would pass today — then the default could later be renamed (main → trunk, say), and that ruleset would silently stop covering the new default with no self-healing mechanism, while an earlier audit run had already reported it clean. That's a strictly weaker configuration validated as equivalent to the one repo-rules actually writes — the same class of quietly-degraded protection this whole PR's audit exists to catch, just one level removed. So the check correctly requires the symbolic token specifically; accepting a literal substitute would reopen the failure mode in scope here.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by mikelward** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_
